autogpt_platform/backend/backend/data/execution.py

```diff
@@ -10,62 +20,144 @@
+    total_run_time: float = Field(..., description="Seconds of node runtime")
+    status: ExecutionStatus
     graph_id: str
-    node_exec_id: str
-    node_id: str
-    block_id: str
-    data: BlockInput
+    graph_version: int
+    preset_id: Optional[str] = None
 
+    @staticmethod
+    def from_db(_graph_exec: AgentGraphExecution):
+        now = datetime.now(timezone.utc)
+        start_time = _graph_exec.startedAt or _graph_exec.createdAt
+        end_time = _graph_exec.updatedAt or now
+        duration = (end_time - start_time).total_seconds()
+        total_run_time = duration
+
+        try:
+            stats = type_utils.convert(_graph_exec.stats or {}, dict[str, Any])
+        except ValueError:
+            stats = {}
+
+        duration = stats.get("walltime", duration)
+        total_run_time = stats.get("nodes_walltime", total_run_time)
+
+        return GraphExecutionMeta(
+            id=_graph_exec.id,
+            user_id=_graph_exec.userId,
+            started_at=start_time,
+            ended_at=end_time,
+            cost=stats.get("cost", None),
+            duration=duration,
+            total_run_time=total_run_time,
+            status=ExecutionStatus(_graph_exec.executionStatus),
+            graph_id=_graph_exec.agentGraphId,
+            graph_version=_graph_exec.agentGraphVersion,
+            preset_id=_graph_exec.agentPresetId,
+        )
 
-ExecutionStatus = AgentExecutionStatus
 
-T = TypeVar("T")
+class GraphExecution(GraphExecutionMeta):
+    inputs: dict[str, Any]
+    outputs: dict[str, list[Any]]
+    node_executions: list["NodeExecutionResult"]
 
+    @staticmethod
+    def from_db(_graph_exec: AgentGraphExecution):
+        from .block import BlockType, get_block
 
-class ExecutionQueue(Generic[T]):
-    """
-    Queue for managing the execution of agents.
-    This will be shared between different processes
-    """
+        if _graph_exec.AgentNodeExecutions is None:
+            raise ValueError("Node executions must be included in query")
 
-    def __init__(self):
-        self.queue = Manager().Queue()
+        graph_exec = GraphExecutionMeta.from_db(_graph_exec)
 
-    def add(self, execution: T) -> T:
-        self.queue.put(execution)
-        return execution
+        node_executions = sorted(
+            [
+                NodeExecutionResult.from_db(ne, _graph_exec.userId)
+                for ne in _graph_exec.AgentNodeExecutions
+            ],
+            key=lambda ne: (ne.queue_time is None, ne.queue_time or ne.add_time),
+        )
 
-    def get(self) -> T:
-        return self.queue.get()
+        inputs = {
+            **{
+                # inputs from Agent Input Blocks
+                exec.input_data["name"]: exec.input_data.get("value")
+                for exec in node_executions
+                if (
+                    (block := get_block(exec.block_id))
+                    and block.block_type == BlockType.INPUT
```

Comment:
nit: can we check "name" in exec.input_data to avoid key error ?

Reply:
Is it even possible to save a graph with nameless inputs or outputs? The `name` field is required on both `AgentInputBlock` and `AgentOutputBlock` so I'd expect it to always be set and therefore an extra check wouldn't be necessary.

Reply:
If we disable the validation during save (not run), to avoid preventing the user saving the graph that is still in progress, then yes.

Reply:
but since this is an execution maybe it doesn't matter